### ISSUE TITLE
Core Files Generated on Arch Linux

### ISSUE THREAD

**Issue by domenukk** [AFLplusplus/AFLplusplus#324]
Right now, every crash in AFL++ creates a coredump. As crashes are reproducible anyway, how about (optionally?) changing the core pattern to `kernel.core_pattern=|/bin/false`?

/edit: This is on arch linux.

**Comment by domenukk** [comment]
It's actually there for a reason, but maybe there are less "loud" ways to achieve it?
https://github.com/google/fuzzer-test-suite/issues/60

**Comment by vanhauser-thc** [comment]
how is this a problem? (with the core_pattern being core?)

**Comment by domenukk** [comment]
I have hundreds of useless crash.1234 files everywhere

**Comment by vanhauser-thc** [comment]
ah so your kernel.core_pattern is not "core" but "core.%p" (or something like that).
so you didnt call afl-system-config or another script change the core_pattern.

I do not have a single core file in my fuzzing targets.

**Comment by domenukk** [comment]
Actually I do use `afl-system-config`. I've seen this happen at least on other Arch Linux systems, so it may be the distrubution's fault.

The wiki specifically names 
`kernel.core_pattern=|/bin/false` 
and 
```
/etc/systemd/coredump.conf
[Coredump]
Storage=none
```

Thinking of it - which problem is not caused by systemd?

**Comment by vanhauser-thc** [comment]
so maybe this needs an update in afl-system-config to systems that need something different. I dont have/use Arch - so if you find out why try to find the workaround :)

**Comment by vanhauser-thc** [comment]
@domenukk any progress on this?

**Comment by domenukk** [comment]
I don't have any core files lying around in my current setup. Chances are this only happens after some update or refresh with afl already running, resetting the core pattern?
Or maybe some other setup changed in arch in the meantime. That stuff definitely happens...
Closing for now until it comes back.

**Comment by dkasak** [comment]
Interestingly, this just started happening to me too after I updated from about a year ago to the latest stable. I'm also on Arch Linux.

```
❯ sysctl kernel/core_pattern 
kernel.core_pattern = core
```

The `-M` instance is not producing any core files, but as soon as I start an `-S` instance, I get loads of them named as `core.NUM`, which exhausts my disk space. The UI of the `-S` instances does not appear to be showing crashes, so this may or may not be the same issue as described above, but I have a feeling it's related at least somehow.

**Comment by domenukk** [comment]
> /etc/systemd/coredump.conf
> [Coredump]
> Storage=none

Have you tried this?
The core dumps may also occur when the target gets killed due to timeouts, I guess.
It's an odd beast, and only seems to affect Arch...

**Comment by dkasak** [comment]
I just did, doesn't help.

There are no timeouts either. The coredumps were created by a SIGABRT and the backtraces are all like this:

```
#0  0x00007ffff7ac4ef5 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7aae862 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff7b06f38 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007ffff7b0ebea in malloc_printerr () from /usr/lib/libc.so.6
#4  0x00007ffff7b1044c in _int_free () from /usr/lib/libc.so.6
#5  0x00007ffff7b13ca8 in free () from /usr/lib/libc.so.6
#6  0x000055555555d4f8 in read_file (fd=0, buffer=0x7fffffffab78) at fuzzers/include/fuzzing.hh:40
#7  0x000055555555d952 in main (argc=<optimized out>, argv=<optimized out>) at fuzzers/fuzz_group_decrypt.cpp:29(gdb) 
```

This is with LTO binaries so I'll have to try without LTO to verify it's not some weirdness related to that. Though it's still a mystery why it's not following the core pattern in the first place...

**Comment by dkasak** [comment]
I recompiled without LTO and the symptoms are the same. Seems I'll have to dig deeper into this.

**Comment by dkasak** [comment]
Well, that's weird... I bisected and it turns out it starts happening on 1efc6e59b7cd2a3623ad3d75622e8bf107b3dc98.

**Comment by dkasak** [comment]
Turns out that reverting 1efc6e59b7cd2a3623ad3d75622e8bf107b3dc98 doesn't help at d0225c2c4d465968660a08c93857fed354e539b1 (current stable), but it does help on some earlier commits. Due to this, I tried bisecting from 1efc6e59b7cd2a3623ad3d75622e8bf107b3dc98 to d0225c2c4d465968660a08c93857fed354e539b1, but reverting 1efc6e59b7cd2a3623ad3d75622e8bf107b3dc98 on each bisection step (with minor differences to resolve merge conflicts).

The revert commit looked something like this:

```diff
diff --git a/src/afl-forkserver.c b/src/afl-forkserver.c
index e59f0d11..905a0aba 100644
--- a/src/afl-forkserver.c
+++ b/src/afl-forkserver.c
@@ -1013,7 +1013,7 @@ void afl_fsrv_write_to_testcase(afl_forkserver_t *fsrv, u8 *buf, size_t len) {
 
     s32 fd = fsrv->out_fd;
 
-    if (!fsrv->use_stdin && fsrv->out_file) {
+    if (fsrv->out_file) {
 
       if (unlikely(fsrv->no_unlink)) {
 
@@ -1044,7 +1044,7 @@ void afl_fsrv_write_to_testcase(afl_forkserver_t *fsrv, u8 *buf, size_t len) {
 
     ck_write(fd, buf, len, fsrv->out_file);
 
-    if (fsrv->use_stdin) {
+    if (!fsrv->out_file) {
 
       if (ftruncate(fd, len)) { PFATAL("ftruncate() failed"); }
       lseek(fd, 0, SEEK_SET);
diff --git a/src/afl-fuzz-init.c b/src/afl-fuzz-init.c
index 56dae48c..5ea22e70 100644
--- a/src/afl-fuzz-init.c
+++ b/src/afl-fuzz-init.c
@@ -2053,26 +2053,24 @@ void setup_cmdline_file(afl_state_t *afl, char **argv) {
 
 void setup_stdio_file(afl_state_t *afl) {
 
+  u8 *fn;
   if (afl->file_extension) {
 
-    afl->fsrv.out_file =
-        alloc_printf("%s/.cur_input.%s", afl->tmp_dir, afl->file_extension);
+    fn = alloc_printf("%s/.cur_input.%s", afl->tmp_dir, afl->file_extension);
 
   } else {
 
-    afl->fsrv.out_file = alloc_printf("%s/.cur_input", afl->tmp_dir);
+    fn = alloc_printf("%s/.cur_input", afl->tmp_dir);
 
   }
 
-  unlink(afl->fsrv.out_file);                              /* Ignore errors */
+  unlink(fn);                                              /* Ignore errors */
 
-  afl->fsrv.out_fd = open(afl->fsrv.out_file, O_RDWR | O_CREAT | O_EXCL, 0600);
+  afl->fsrv.out_fd = open(fn, O_RDWR | O_CREAT | O_EXCL, 0600);
 
-  if (afl->fsrv.out_fd < 0) {
+  if (afl->fsrv.out_fd < 0) { PFATAL("Unable to create '%s'", fn); }
 
-    PFATAL("Unable to create '%s'", afl->fsrv.out_file);
-
-  }
+  ck_free(fn);
 
 }
```

Using this technique, 981ffb27a8a166b51a06d57fce044ed1eaf1aa62 happens to be the first commit where reverting 1efc6e59b7cd2a3623ad3d75622e8bf107b3dc98 doesn't fix the issue anymore.

I'm not sure whether this brings us any closer to solving the issue or is some kind of a red herring due to something else.

**Comment by vanhauser-thc** [comment]
@dkasak I dont understand how -M vs -S or that commit vs before changes if core files are generated for found crashes or not.
makes no sense at all ...

**Comment by dkasak** [comment]
It makes no sense to me either. I just decided to report the raw observables while I'm thinking about it. I suppose there's something else going on that I'm missing.

Note that these *don't* appear to be *found* crashes. The afl++ UI doesn't show it found either crashes or timeouts, but core files are still getting created.

**Comment by vanhauser-thc** [comment]
oh that is weird. well what is the output if you run "file core…."?

**Comment by dkasak** [comment]
core.1200237: ELF 64-bit LSB core file, x86-64, version 1 (SYSV), SVR4-style, from 'build/fuzzers/fuzz_group session.txt', real uid: 1000, effective uid: 1000, real gid: 1000, effective gid: 1000, execfn: 'build/fuzzers/fuzz_group', platform: 'x86_64'

Why? It *is* a valid core file, just in case you missed https://github.com/AFLplusplus/AFLplusplus/issues/324#issuecomment-832775448.

**Comment by dkasak** [comment]
I also just tried running with `AFL_DEBUG_CHILD=1` and noticed a bunch of `double free or corruption (top)` which seem to appear with the moments that core files are appearing in.

**Comment by vanhauser-thc** [comment]
@dkasak can you share your setup? like what you are fuzzing (link) how you instrument the target (exact command line with envs your set) and how you run the target (out/default/target_setup).

**Comment by hexcoder-** [comment]
@dkasak Is your setup [olm](https://gitlab.matrix.org/matrix-org/olm/-/tree/master/)?

**Comment by dkasak** [comment]
Yes, but it turns out the target is irrelevant. I figured out that the fuzzing shim has a bug which sometimes causes a double free and this turns out to be the root cause of this issue.

I managed to create a minimal reproducing test case as follows.

Save the following as `fuzzer.c`:

```c
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <unistd.h>
#include <string.h>

int main(void) {
    uint8_t buffer[4096];

    read(STDIN_FILENO, &buffer, 4096);

    int len = strlen((char *) buffer);

    if (len == 42) {
        uint8_t *buf = malloc(4096);
        realloc(buf, 0);
        free(buf);
    }

    return 0;
}
```

Compile `fuzzer.c` with

    afl++/afl-clang-fast -o fuzzer fuzzer.c

Then run

    mkdir in out
    printf %100s > in/1
    afl++/afl-fuzz -i in -o out -S s -- ./fuzzer

This will start creating core files very soon, every time it manages to hit a file which is a C string of length 42.

The necessary condition seems to be to free once with `realloc(..., 0)`, once with `free`. A double free by calling `free` twice won't trigger it.

In the original fuzzing shim, this would only happen when the length of the file was 0, so it would only be hit once the trimming algorithm trimmed the file down to 0 bytes.

**Comment by vanhauser-thc** [comment]
@dkasak the harness example is obliviously wrong. realloc has return value that has to be used.

ok, so a core file is generated when the main fuzzing process crashes and not when a crash is triggered in a child. IMHO this is ok, because this means something is wrong in the setup or harness, but this should still not happen ... weird. 
will try to reproduce here on my debian

**Comment by vanhauser-thc** [comment]
ah sorry I wrote nonsense, of course this happens in the child, and a crash there should not create a core.

I tested your example, and as it should, on my system no cores are generated.
It is an Arch Linux thing it seems. no need to analyze afl or your harness, you have to check what has to be done on Arch to not have cores.

**Comment by dkasak** [comment]
> the harness example is obliviously wrong. realloc has return value that has to be used.

Yes, it's blatantly wrong on purpose, since I wanted to demonstrate the minimal example that would produce core files. The original harness had the same bug but in a subtler way.

It seems you're right; calling the harness manually on the appropriate input generates a core file as well. It's just weird since I've never had the problem of proliferating core files before I updated afl++ so I assumed it must've been something in there that changed. What's also weird is that the original harness wasn't registering these as crashes in the UI, but that may have just been another bug.

Regarding core files on Arch Linux, I think I found an incantation that makes them go away. If you run

    sysctl -w kernel.core_uses_pid=0

That makes the `.PID` part go away, so that your core pattern is used literally.

Setting an empty core pattern seems to make them go away entirely

    sysctl -w kernel.core_pattern=""
